streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1940,111 +1940,120 @@ object ZChannel {
-                                 _    <- ZIO.when(size >= n)(cancelers.take.flatMap(_.succeed(())))
-                                 _    <- cancelers.offer(canceler)
-                                 raceIOs =
-                                   ZIO.scopedWith { scope =>
-                                     (incoming >>> channel)
-                                       .toPullInAlt(scope)
-                                       .flatMap(evaluatePull(_).race(canceler.await.interruptible))
-                                   }
-                                 _ <- permits
-                                        .withPermit(latch.succeed(()) *> raceIOs)
-                                        .interruptible
-                                        .forkIn(childScope)
-                                 _ <- latch.await
-                               } yield ()
-                             }
-                       }
-        _ <- pullStrategy(fiberId).forever
-               .catchAllCause(cause =>
-                 cause.failureOrCause match {
-                   case Left(_: Left[OutErr, OutDone]) => outgoing.offer(Exit.failCause(cause))
-                   case Left(x: Right[OutErr, OutDone]) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *>
-                       lastDone.get.flatMap {
-                         case Some(lastDone) => outgoing.offer(Exit.fail(Right(f(lastDone, x.value))))
-                         case None           => outgoing.offer(Exit.fail(Right(x.value)))
-                       }
-                   case Right(cause) => outgoing.offer(Exit.failCause(cause.map(Left(_))))
-                 }
-               )
-               .raceFirst(awaitErrorSignal(childScope, fiberId)(errorSignal))
-               .forkIn(scope)
-      } yield {
-        lazy val consumer: ZChannel[Env, Any, Any, Any, OutErr, OutElem, OutDone] =
-          unwrap[Env, Any, Any, Any, OutErr, OutElem, OutDone] {
-            outgoing.take.flatten.foldCause(
-              cause =>
+      ZIO.fiberIdWith { fiberId =>
+        ZIO.suspendSucceed {
+          val input          = SingleProducerAsyncInput.unsafe.make[InErr, InElem, InDone](fiberId)
+          val incoming       = ZChannel.fromInput(input)
+          val n0             = n.toLong
+          val bufferSize0    = bufferSize
+          val mergeStrategy0 = mergeStrategy
+          val outgoing =
+            Queue.unsafe.bounded[ZIO[Env, Either[OutErr, OutDone], OutElem]](bufferSize0, fiberId)(Unsafe.unsafe)
+          val cancelers   = Queue.unsafe.unbounded[Promise[Nothing, Unit]](fiberId)(Unsafe.unsafe)
+          val lastDone    = Ref.unsafe.make[OutDone](null.asInstanceOf[OutDone])(Unsafe.unsafe)
+          val errorSignal = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe.unsafe)
+          val permits     = Semaphore.unsafe.make(n0)(Unsafe.unsafe)
+
+          type Pull[A] = ZIO[Env, Either[OutErr, OutDone], A]
```

Comment:
As I was writing this, I felt like the `toPull` stuff should pull chunks instead of single elements. The cost is a little bit of housekeeping for error handling. The chunk is already somewhere in memory for the by-element pull, so it makes no difference for correctness, however it makes a huge difference for performance.

Reply:
I'm a bit busy right now. I'll explore your idea later. Thanks for your review! 🙂

Reply:
I had a look, I think to have any real effect we will need chunk-aware channels and update the pipelines:
https://github.com/zio/zio/blob/2b4f239b0531a1e5c05b4cf97e8cb995f3637e22/streams/shared/src/main/scala/zio/stream/ZPipeline.scala#L1839C5-L1845C6
to something like:
```
new ZPipeline(
      ZChannel
        .identity[Nothing, Chunk[In], Any]
        .mapOutChunkZIOPar(n, bufferSize)(f)
    )
```
with the caveat that it breaks chunk structure.
```
final def mapOutZIOPar[Env1 <: Env, OutErr1 >: OutErr, InElem0, OutElem0, OutElem2](n: Int, bufferSize: Int = 16)(
      f: OutElem0 => ZIO[Env1, OutErr1, OutElem2]
)(implicit
      ev0: Chunk[InElem0] <:< InElem,
      ev1: OutElem <:< Chunk[OutElem0],
      trace: Trace
): ZChannel[Env1, InErr, Chunk[InElem0], InDone, OutErr1, Chunk[OutElem2], OutDone] = ???
```